### PR TITLE
Add support for multi node farmers

### PR DESCRIPTION
**Pull request by felixbrucker**
### Purpose:
This PR allows the farmer to be configured with multiple full nodes to be able to continue farming while n-1 nodes are unavailable/down. This can be useful when currently using a single remote full node hosted locally or in the cloud, where having another full node helps to reduce farm downtime and making maintenance easier.

### Current Behavior:

Currently when configuring a farmer to use multiple full nodes it processes the new signage points multiple times, once for each full node. Additionally the signed values required in the block forging process are sent to all nodes, not just the node which originally requested them, resulting in n*n `SignedValues` messages

### New Behavior:

This PR fixes those two issues:
- The farmer now checks for duplicate SPs before processing the new SP to avoid duplicate lookups
- The `request_signed_values` api method of the farmer now handles its state by using `call_api` to keep track of the harvester response and returns the appropriate `SignedValues` message to the full node itself

### Testing Notes:

I added tests regarding the duplicate SP processing, returning the signed values from the farmer and the `call_api_of_specific` helper.

### CODE REVIEW

**Comment by jack60612** [comment]
any way for you to make the diff on farmer_api,py not look so messy?

**Comment by felixbrucker** [comment]
@jack60612 i have split up the commit into smaller chunks, the first commit only moves logic into a private method, but git still makes a mess there, don't think that can be avoided, but the change is much smaller now. All other commits should be easier to read now 👍

**Comment by jack60612** [comment]
> @jack60612 i have split up the commit into smaller chunks, the first commit only moves logic into a private method, but git still makes a mess there, don't think that can be avoided, but the change is much smaller now. All other commits should be easier to read now 👍

Sounds good & sorry for the delay and feel free to bug me on discord / gh in the future.

**Comment by jack60612** [comment]
I'll review this in the morning.

File: chia/farmer/farmer_api.py
Comment (jack60612) [inline chia/farmer/farmer_api.py:607]: not sure if this None is needed.

File: tests/connection_utils.py
Comment (jack60612) [inline tests/connection_utils.py:57]: could you more clearly define this is a set

File: chia/harvester/harvester_api.py
Comment (jack60612) [inline chia/harvester/harvester_api.py:269]: good catch (feel free to dismiss this)

File: chia/farmer/farmer_api.py
Comment (jack60612) [inline chia/farmer/farmer_api.py:365]: How about simply adding it to the list right here, and then not having to worry about an exception?
I'm not too worried about something getting added twice as the processing should be pretty fast, if you disagree let me know.

File: chia/farmer/farmer_api.py
Comment (jack60612) [inline chia/farmer/farmer_api.py:407]: maybe adding an additional comment explaining why you separated this function would be nice.

**Review by jack60612 — CHANGES_REQUESTED** [review]
feel free to do the merge now.

**Comment by jack60612** [comment]
oh I just realized all those comments weren't one review, sorry about that.

File: chia/farmer/farmer_api.py
Comment (felixbrucker) [reply chia/farmer/farmer_api.py:607]: It's not needed, but recommended for functions which actually return stuff (pep8), if desired i can remove it ofc

Comment (jack60612) [reply chia/farmer/farmer_api.py:607]: should be fine then, was not sure if mypy cared or not.

File: chia/farmer/farmer_api.py
Comment (felixbrucker) [reply chia/farmer/farmer_api.py:365]: Duplicate SP processing by the harvesters within a very short timeframe will cause multiple reads and processing in case of compressed plots, which is undesired

Imho ensuring we do not do unnecessary reads and compute is important, is it not?

Comment (jack60612) [reply chia/farmer/farmer_api.py:365]: fair point with the compressed plots.

**Comment by jack60612** [comment]
is the ci good?

**Comment by felixbrucker** [comment]
It should go through except for the coverage one which complains about the extracted processing, as the farmer_api didn't have any tests yet

**Review by jack60612 — DISMISSED** [review]
lgtm, I'm going to get @Chia-Network/required-reviewers approval now.

**Comment by jack60612** [comment]
@felixbrucker we are going to review this after the 19th.

**Comment by jack60612** [comment]
@felixbrucker don't worry about resolving the conflicts, ill let you know when we are going to review it.

**Comment by jack60612** [comment]
@felixbrucker you are good to resolve now

**Review by jack60612 — APPROVED** [review]
lgtm

**Review by wjblanke — APPROVED** [review]
aok

**Pull request merged**
